### PR TITLE
Add 4th order S_CURVE_ACCELERATION with configurable S_CURVE_FACTOR

### PR DESCRIPTION
**Pull request by mh-dm**
### Description

Add configurable 4th order motion support within `S_CURVE_ACCELERATION` alongside the existing 6th order motion support.

### Background

4th? 6th? What? Short explanation: by default (trapezoidal motion), you set an acceleration limit (example 500mm/s^2), and then whenever the gcode/instructions involve a change in speed it is achieved by accelerating (or decelerating) at that set limit for the required amount of time to achieve the desired speed change. The acceleration goes from 0 (at previous speed) ~instantly to 1x of set limit then back to 0 (at newly desired speed). Force is mass times acceleration which means sudden changes in forces whenever changing speed.

s-curve-no
Trapezoidal motion, showing only X-axis. Most speed changes are at 500mm/s^2 acceleration limit.

However, with `S_CURVE_ACCELERATION` the acceleration will smoothly ramp up from 0, go to roughly 2x of the limit then smoothly ramp down to 0 exactly right to achieve the desired speed change. 2x? Yes. We must temporarily go over 1x to complete the same speed change in the same amount of time but while ramping from 0 acceleration. Overall, the average acceleration remains at the set acceleration limit.

s-curve-6th
`S_CURVE_ACCELERATION` 6th order, showing reaching roughly 2x of the acceleration limit

It's "6th order" because the motion model is position over time as a 6th order polynomial function.
However, the actual implementation works with velocity (derivative of position) modeled as a 5th order polynomial.

### Details

The problem, I might argue, is that it's "too smooth" for some applications, including 3d printers. `S_CURVE_ACCELERATION` 6th order spends a lot of time around 0 slowly ramping up then it has to hit high acceleration levels. The solution would be a model with a lower order polynomial. So I did the math for a 4th order polynomial (3rd order in velocity) and this is the result:

s-curve-4th
4th order, with `S_CURVE_FACTOR` of 0

Note how it only reaches roughly 1.6x of the acceleration limit.

The 4th order motion allows some configuration: the acceleration will start at `S_CURVE_FACTOR` of the limit, ramp up to above 1x then smoothly back down to `S_CURVE_FACTOR` of the limit exactly right to achieve the desired speed change.

s-curve-4th-0 25
4th order, with `S_CURVE_FACTOR` of 0.25

Note how it only reaches roughly 1.4x of the acceleration limit.

### Benefits

Less smooth is more smooth™.

For 3d printing in particular, parts are printed line by line with direction changes/corners in between. Cornering is done with "jerk" or junction deviation, both of which involve basically instant speed changes. Basically instant speed change means high acceleration spikes. That doesn't pair that well with a very slowly smoothly ramping `S_CURVE_ACCELERATION`. It should pair better with a still smooth 4th order motion model that ramps up faster, like starting at a 0.25 `S_CURVE_FACTOR`. Lower peak acceleration should pair better with low-torque motors used in 3d printers.

### TL;DR

Enable `S_CURVE_ACCELERATION`, uncomment `S_CURVE_FACTOR 0.25` and check whether you get **less ringing** and/or if the printer sounds better/quieter. Then check how much you can **increase the acceleration** limits before you get too much ringing.

_Acceleration only_, this will have zero or minimal effect on ringing induced by jerk/junction deviation (but you don't know which type of ringing you have until you test). For the other kind of ringing I think you can try input shaping.

### Configurations

**This feature is not implemented for AVR**. If interested I encourage you to have a go at implementing if you have a logic analyzer, an AVR board and are interested in assembly.

The graphs were obtained with a logic analyzer capturing 15M samples at 12Mhz, directly from the X-axis step/dir pins. (Just FYI my boards are LPC176x). Tiny note that captures were done with PRs #26881 and #27035 already applied. I did that to fix some issues that would pollute the captures a tiny bit.

Test gcode:

```gcode
G4 P550 ; Delay for capture
G92 X10 Y10
M92 X80 Y80 ; 80 steps per mm
M204 S500
M205 X0.2 Y0.2 ; Low jerk
G0 F1500 X11.5
G0 F1200 X12
G0 F1800 X13.5
G0 F1400 X14.1
G0 F1600 X15.2
G92 Y10 ; sync test
G0 F1800 Y10.1
M204 S250
G0 F2000 X15.33
M204 S350
G0 X15.55
M204 S500
G0 X15.85
M204 S750
G0 X16.5
M204 S420
G0 X16.77
M204 S250
G0 X17
G92 Y10 ; sync test
M204 S500
G0 Y10.1
G0 F600 X17.4
G0 F1800 X18.95
G0 F600 X19.3
G0 F1800 X20.9
G0 F600 X21.3
G0 F1800 X23
G0 F600 X23.4
G0 F1800 X25.2
G0 F1500 X26
G0 F1200 X27
G0 F1800 X24.5
G92 Y10 ; sync test
G0 Y10.1
G0 F6000 X10
```

### CODE REVIEW

**Comment by InsanityAutomation** [comment]
Nice work again! Configurable bow was the biggest drawback to s-curve in my opinion! One more thing to pull into my test branch! Lol

**Comment by cbagwell** [comment]
This will be interesting to test on boards with an STM32G0 (such as SKR Mini E3 V3.0 and BTT Manta boards) because they are using the fallback implementation for S_CURVE_ACCELRATION and that section runs in 500 cycles on those boards since it doesn't have the umull instruction.   Each multiplication on STM32G0 takes 52 cycles so this PR takes around 100 cycles off that 500.

**Comment by vovodroid** [comment]
But there is a question - does S_CURVE have meaning at all, taking in consideration that jerk, happening just before/after acceleration/deceleration produces infinite (obviously not) acceleration?

**Comment by mh-dm** [comment]
@vovodroid Yes S_CURVE still has meaning but in different ways to different people:
1) if the primary objective is smooth motion you can use S_CURVE and just set jerk low enough. Note that it's not actually infinite acceleration because the jerk (cornering change in speed) happens from one step to the next so over the duration of a step. At 10mm/s cornering speed and 80 steps/mm that's 800 steps/s or 0.00125s for one step. So a 1mm/s jerk then is equivalent to 800mm/s^2 acceleration. People can do the math or lower jerk and test until it works for their application.
2) if the primary objective is fast motion/printing then S_CURVE can work with jerk motion. A high jerk means a really high momentary acceleration at the corner. Without S_CURVE you also usually have full deceleration on the segment up to the corner then full acceleration after. S_CURVE avoids all (or most with S_CURVE_FACTOR) of that, flattening the max load on the system. In theory this means less ringing than without S_CURVE at high acceleration and jerk. In practice, the improvement I've personally seen is small. However, without more specific testing/calibration it's hard to judge since the improvement is highly sensitive to the machine's resonance characteristics and the acceleration used. (If you want to know why it's highly sensitive, there's this cool explainer but note that it doesn't reference Marlin 'jerk' rather it uses is 'physics jerk', which I think should be referred to as __jolt__ in 3d printing contexts: https://xenovacivus.github.io/s-vs-t-curve/s-vs-t-curve.html by @xenovacivus . It also confusingly changes the actual overall acceleration of the motion when changing the physics jerk but the overall idea is there)

**Comment by dbuezas** [comment]
@mh-dm have you confirmed this modification improves print quality? Do you have some pictures?

**Comment by vovodroid** [comment]
Hi, is there ready to use formula to get peak acceleration from S_CURVE_FACTOR?

**Comment by mh-dm** [comment]
@dbuezas As noted in my August comment: "In theory this means less ringing than without S_CURVE at high acceleration and jerk. In practice, the improvement I've personally seen is small". I saw a small improvement on ringing on some test prints but I didn't take proper photos. I'm doing a round of testing in the next couple weeks for other reasons (swapping belts) so I can compare without vs 0.25 S_CURVE_FACTOR vs pre-existing S_CURVE.

@vovodroid Deriving the formula would probably be an interesting math exercise but I haven't done it. From the data I've shown above: 0, 0.25, and 1 S_CURVE_FACTOR results in momentary peak accelerations of roughly 1.6x, 1.4x, and 1x the set acceleration limit respectively. Best guess it's close enough to a linear interpolation between 1.6x and 1x.

**Comment by vovodroid** [comment]
Can S_CURVE (both current and 4th) go over absolute limits set by M201?

**Comment by mh-dm** [comment]
S_CURVE (current and this PR) is implemented such that it makes the required speed changes in the exact same amount of time that trapezoid motion does. In other words the average acceleration is the same as for trapezoid motion and thus we can say that, _on average_, it respects the M201 configuration.

But to more directly answer your question: Yes, for peak acceleration. Current S_CURVE peaks at ~2x M201 limits and this PR proposes 4th order S_CURVE that peaks at significantly lower, 1.6x or less depending on S_CURVE_FACTOR.

Good question as this means: The higher peak acceleration means that switching to current S_CURVE has a significant potential to result in skipped steps if the acceleration configuration is set very close to machine limits. It's not as bad as the ~2x implies since A) peak acceleration is different (and 'easier') than average acceleration and B) the peak acceleration happens at a different time than when direction changes / jerk happen (which demand really high peak accelerations from the machine, also past the M201 acceleration configuration).

**Comment by mh-dm** [comment]
As promised here's a side by side comparison between the 3 versions. To reduce bias I will reveal which is which at the end of this post.

1-wall prints that are lying on the side (for even lighting and landscape). Bottom/bed side is on the left. The print was also a testing tower in that the XY jerk values increased every layer from 10 at the bottom to 40 at the top (example `M205 X33.40 Y33.40`).

Face parallel to X axis:
parallel-to-x
Opposite face:
opposite-x
Face parallel to Y axis:
parallel-to-y
Opposite face
opposite-y

The photos are closeups - from a bit further away and less than perfect light only one 'ring' is actually noticeable. Hard to take photos as they were either blurry or had way too much moire. I think the random splotches on the round surfaces are due to filament being a bit wet. Printing from dry box, currently at 31% humidity so I guess I need to recharge the silica gel..

Always printed the same gcode, just the firmware configuration on the printer was different. Print settings: 0.2 layer height, 0.4 nozzle, 0.5 layer width, 100mm/s print speed. Adv. K 0.15. A more aggressive 6000 print acceleration which I made sure was the real limiting factor. 1) printer limits were above: `M201 X12000.00 Y6000.00 Z500.00 E700.00` 2) In marlin `planner.cpp` I kinda disabled the `max_accel_steps_per_s2` limit that slows down acceleration based on max e jerk. Yes that implies my E stepper was commanded to the equivalent of 6000×0.15×0.2×0.5÷(pi×1.75^2÷4) ~= 37.4 of E jerk, which is ok on a bowden with a big E stepper motor.

Printer is a literally cut down ender-3. 180x180 print bed so significantly less moving mass on the Y-axis which explains why X and Y are so similar and partly explains why there's not much ringing at ~35 mm/s Y jerk. What's a bit counterintuitive is that at lower jerk values there's just slightly more ringing - hard to explain but I think a large component of the ringing may be from the bowden/extruder decelerating/accelerating but I'm getting too off-topic. I almost forgot to reveal the order: s_curve (6th) then trapezoid then s_curve_factor 0.25. At least on my machine and from what I can tell in person trapezoid and 4th order are similar with a slight edge for the 4th order and both are better than 6th order.

**Comment by dbuezas** [comment]
I think far right looks best. Do you also have input shaping enabled?
I'd ask for one with SmoothLA (which also tracks pressure through the S curve), but that shouldn't affect beyond the first ringing

**Comment by mh-dm** [comment]
Input shaping was not enabled. Sure, will do SmoothLA esp. since I wanted to try it out, just need to sync/rebase so I'll take some time.

**Comment by dbuezas** [comment]
Just remembered s-curve support is not merged yet. The PR is simple though (draft because code is quick and dirty, it works fine).

**Comment by vovodroid** [comment]
@mh-dm hi, any chances that this feature can cause such consequence layer shifts to one side?

image

**Comment by mh-dm** [comment]
The existing/default S_CURVE can make layer shifts more likely because the peak acceleration is ~2x the normal acceleration. The proposed 4th order S_CURVE (ex. with S_CURVE_FACTOR of 0.25) results in reduced peak acceleration of 'only' roughly ~1.4x.
I've been running with S_CURVE_FACTOR 0.25 (or 0.2) since before I've made this PR and if I remember correctly the only layer shift I've experienced was minor and a build plate clips issue (2 out of 4 clips.. surprised it finished). Partly due to the exturder, my acceleration limits (5k mm/s^2 for y) are not close to machine limits: 10-12k mm/s^2 in simple y back and forth move tests. From same simple move tests on my modified full size ender-3, the effect of having s_curve on or not on maximum achievable acceleration is minimal (example: 16k no_s_curve vs 15k for s_curve for travels done at 400mm/s but no difference if moving is done at 500/600/650/800/1000). Steppers (driver&motor) seem to handle well the peak in s_curve acceleration.

I'd look for a hardware cause (like the problematic axis having too much friction or not moving). Though if you're running with >5k mm/s maybe the tune is too aggresive for the _moving mass_ of the problematic axis?

**Comment by vovodroid** [comment]
Well, I disabled `S_CURVE_FACTOR`, since that I had several (may be 4-5) 5-6 hours prints with and without reboot, and one 12 hours, without any problems. All other factors remain the same, including 28-29C ambient temperature.

I still think about some software issue. It's difficult to suggest mechanical or electronic failure producing such consistent  layer shift.

Could there be some rounding or overflow is 4th order implementation?

image

**Comment by vovodroid** [comment]
Another successful 8 hours print without S_CURVE_FACTOR

image
